weather_dl/download_pipeline/pipeline.py

```diff
@@ -150,108 +93,112 @@ def skip_partition(config: t.Dict, store: Store) -> bool:
+      [parameters.bob]
+      api_key=KKKKK2
+      api_url=UUUUU2
+      [parameters.eve]
+      api_key=KKKKK3
+      api_url=UUUUU3
+    ```
+    """
+    return [(name, params) for name, params in config['parameters'].items()
+            if isinstance(params, dict)] or [('default', {})]
+
+
+def prepare_partitions(config: Config, store: t.Optional[Store] = None) -> t.Iterator[Partition]:
+    """Iterate over client parameters, partitioning over `partition_keys`.
+
+    First, this produces a Cartesian-Cross over the range of keys.
+    For example, if the keys were 'year' and 'month', it would produce
+    an iterable like: ( ('2020', '01'), ('2020', '02'), ('2020', '03'), ...)
+
+    If the `parameters` section contains subsections (e.g. '[parameters.1]',
+    '[parameters.2]'), collect a repeating cycle of the subsection key-value
+    pairs. Otherwise, store empty dictionaries.
+
+    This is useful for specifying multiple API keys for your configuration.
+
+    For example:
+    ```
+      [parameters.alice]
+      api_key=KKKKK1
+      api_url=UUUUU1
+      [parameters.bob]
+      api_key=KKKKK2
+      api_url=UUUUU2
+      [parameters.eve]
+      api_key=KKKKK3
+      api_url=UUUUU3
+    ```
+    """
     if store is None:
         store = FSStore()
-    partition_keys = config['parameters']['partition_keys']
+    partition_keys = config.get('parameters', {}).get('partition_keys', [])
     selection = config.get('selection', {})
 
-    # Produce a Cartesian-Cross over the range of keys.
-    # For example, if the keys were 'year' and 'month', it would produce
-    # an iterable like: ( ('2020', '01'), ('2020', '02'), ('2020', '03'), ...)
     fan_out = itertools.product(*[selection[key] for key in partition_keys])
 
-    # If the `parameters` section contains subsections (e.g. '[parameters.1]',
-    # '[parameters.2]'), collect a repeating cycle of the subsection key-value
-    # pairs. Otherwise, store empty dictionaries.
-    #
-    # This is useful for specifying multiple API keys for your configuration.
-    # For example:
-    # ```
-    #   [parameters.deepmind]
-    #   api_key=KKKKK1
-    #   api_url=UUUUU1
-    #   [parameters.research]
-    #   api_key=KKKKK2
-    #   api_url=UUUUU2
-    #   [parameters.cloud]
-    #   api_key=KKKKK3
-    #   api_url=UUUUU3
-    # ```
-    extra_params = [params for _, params in config['parameters'].items() if isinstance(params, dict)]
-    params_loop = itertools.cycle(extra_params) if extra_params else itertools.repeat({})
-
-    def new_downloads_only(candidate: t.Dict) -> bool:
-        """Predicate function to skip already downloaded partitions."""
-        return not skip_partition(candidate, store)
-
-    return zip(
-        filter(new_downloads_only, [_create_partition_config(option, config) for option in fan_out]),
-        params_loop
+    params_loop = itertools.cycle(get_subsections(config))
+
+    partition_configs = filter(
+        lambda it: new_downloads_only(it, store),
```

Comment:
Why filter here? If I'm reading this correctly, you're actually hitting the storage here to decide on what to filter. Why not do that in its own DoFn so that it can run in parallel? This could be many orders of magnitude slower than the rest of this method. It seems like just generating all possible paths is what you'd want here with a separate filter step that can do whatever test is required to see if you want to actually download later.

Reply:
A little context: This filter is here because there was a bug we found earlier. If we skipped some downloads in a beam.Filter step, there would be a problem where subsections (licenses) wouldn't be evenly distributed to each non-skipped download. 

It occurs to me: I could probably perform the cycle step as a separate `Map` operation that occurs after a `Filter`, and keep these steps _really_ simple.